pmd-core/src/main/java/net/sourceforge/pmd/PmdAnalysis.java

```diff
@@ -100,49 +110,112 @@ private PmdAnalysis(PMDConfiguration config) {
-        final RuleSets ruleSets = RulesetsFactoryUtils.getRuleSetsWithBenchmark(config.getRuleSets(), ruleSetLoader.toFactory());
-        if (ruleSets != null) {
-            for (RuleSet ruleSet : ruleSets.getAllRuleSets()) {
-                builder.addRuleSet(ruleSet);
-            }
+        if (!config.getRuleSetPaths().isEmpty()) {
+            final RuleSetLoader ruleSetLoader = pmd.newRuleSetLoader();
+            final List<RuleSet> ruleSets = ruleSetLoader.loadRuleSetsWithoutException(config.getRuleSetPaths());
+            pmd.addRuleSets(ruleSets);
         }
+        return pmd;
+    }
 
-        return builder;
+    // test only
+    List<RuleSet> rulesets() {
+        return ruleSets;
     }
 
-    @InternalApi
-    static PmdAnalysis createWithoutCollectingFiles(PMDConfiguration config) {
-        return new PmdAnalysis(config);
+    // test only
+    List<Renderer> renderers() {
+        return renderers;
     }
 
+
     /**
      * Returns the file collector for the analysed sources.
      */
     public FileCollector files() {
         return collector; // todo user can close collector programmatically
     }
 
+    /**
+     * Returns a new ruleset loader, which can be used to create new
+     * rulesets (add them then with {@link #addRuleSet(RuleSet)}).
+     *
+     * <pre>{@code
+     * try (PmdAnalysis pmd = create(config)) {
+     *     pmd.addRuleSet(pmd.newRuleSetLoader().loadFromResource("custom-ruleset.xml"));
+     * }
+     * }</pre>
+     */
+    public RuleSetLoader newRuleSetLoader() {
+        RuleSetLoader loader = RuleSetLoader.fromPmdConfig(configuration);
+        loader.setReporter(this.reporter);
+        return loader;
+    }
+
     /**
      * Add a new renderer. The given renderer must not already be started,
      * it will be started by {@link #performAnalysis()}.
      *
      * @throws NullPointerException If the parameter is null
      */
     public void addRenderer(Renderer renderer) {
-        this.renderers.add(Objects.requireNonNull(renderer));
+        AssertionUtil.requireParamNotNull("renderer", renderer);
+        this.renderers.add(renderer);
+    }
+
+    /**
+     * Add several renderers at once.
+     *
+     * @throws NullPointerException If the parameter is null, or any of its items is null.
+     */
+    public void addRenderers(Collection<Renderer> renderers) {
+        renderers.forEach(this::addRenderer);
+    }
+
+    /**
+     * Add a new listener. As per the contract of {@link GlobalAnalysisListener},
+     * this object must be ready for interaction. However, nothing will
+     * be done with the listener until {@link #performAnalysis()} is called.
+     * The listener will be closed by {@link #performAnalysis()}, or
+     * {@link #close()}, whichever happens first.
+     *
+     * @throws NullPointerException If the parameter is null
+     */
+    public void addListener(GlobalAnalysisListener listener) {
```

Comment:
ok, so here landed GlobalAnalysisListener - you can add a listener like a renderer...